### ISSUE TITLE
Balena Preload Fails

### ISSUE THREAD

**Issue by HenrikWittemeier** [balena-io/balena-cli#2249]
# Expected Behavior
`balena-cli/balena preload balena-cloud-raspberrypi3-2.72.0+rev1-v12.3.5.img --debug`
Should preload an image with application.

# Actual Behavior
Fails with Error Message.

Tested on Ubuntu 18.04.5 and 20.10.(Same behaviour)
```
balena-cli/balena preload balena-cloud-raspberrypi3-2.72.0+rev1-v12.3.5.img --debug
[debug] new argv=[/home/henrik/balena-cli/balena,/snapshot/versioned-source/bin/balena,preload,balena-cloud-raspberrypi3-2.72.0+rev1-v12.3.5.img] length=4
Building Docker preloader image. [===                     ] 12%
Step 1/7 : FROM docker:17.12.0-ce-dind
Building Docker preloader image. [======                  ] 25%
Step 2/7 : RUN apk update && apk add --no-cache python3 parted btrfs-progs util-linux sfdisk file coreutils sgdisk
 ---> Using cache
Building Docker preloader image. [=========               ] 37%
Step 3/7 : COPY ./requirements.txt /tmp/
 ---> Using cache
Building Docker preloader image. [============            ] 50%
Step 4/7 : RUN pip3 install -r /tmp/requirements.txt
 ---> Using cache
Building Docker preloader image. [===============         ] 62%
Step 5/7 : COPY ./src /usr/src/app
 ---> Using cache
Building Docker preloader image. [==================      ] 75%
Step 6/7 : WORKDIR /usr/src/app
 ---> Using cache
Building Docker preloader image. [=====================   ] 87%
Step 7/7 : CMD ["python3", "/usr/src/app/preload.py"]
 ---> Using cache
 ---> 20fab6b285bd
Successfully built 20fab6b285bd
Building Docker preloader image. [========================] 100%
| Checking that the image is a writable file
| Finding a free tcp port and getting balena settings
| Checking if the image is an edison zip archive
/ Creating preloader container
\ Starting preloader container
\ Reading image informationWaiting for Docker to start...
| Reading image informationDocker started
| Reading image information
? Select an application RaspberryPi3
? Select a release current
/ Fetching application 1
- Estimating required additional space
- Cleaning up temporary files
manifest.manifest.layers is not iterable

**TypeError: manifest.manifest.layers is not iterable**
    at Preloader._getLayersSizes (/snapshot/versioned-source/node_modules/balena-preload/build/preload.js:455:51)
    at Preloader._getApplicationSize (/snapshot/versioned-source/node_modules/balena-preload/build/preload.js:473:40)
    at runNextTicks (internal/process/task_queues.js:62:5)
    at processImmediate (internal/timers.js:429:9)
    at process.topLevelDomainCallback (domain.js:137:15)
    at async Preloader._getSize (/snapshot/versioned-source/node_modules/balena-preload/build/preload.js:491:16)
    at async Preloader._getRequiredAdditionalSpace (/snapshot/versioned-source/node_modules/balena-preload/build/preload.js:497:34)
From previous event:
    at processImmediate (internal/timers.js:456:21)
    at process.topLevelDomainCallback (domain.js:137:15)
From previous event:
    at Preloader.preload (/snapshot/versioned-source/node_modules/balena-preload/build/preload.js:746:14)
    at PreloadCmd.prepareAndPreload (/snapshot/versioned-source/build/commands/preload.js:305:25)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async PreloadCmd.run (/snapshot/versioned-source/build/commands/preload.js:118:13)
    at async PreloadCmd._run (/snapshot/versioned-source/node_modules/@oclif/command/lib/command.js:43:20)
    at async Config.runCommand (/snapshot/versioned-source/node_modules/@oclif/config/lib/config.js:175:24)
    at async CustomMain.run (/snapshot/versioned-source/node_modules/@oclif/command/lib/main.js:27:9)
    at async CustomMain._run (/snapshot/versioned-source/node_modules/@oclif/command/lib/command.js:43:20)
    at async Promise.all (index 1)
    at async oclifRun (/snapshot/versioned-source/build/app.js:75:5)
    at async Object.run (/snapshot/versioned-source/build/app.js:88:9)

For further help or support, visit:
https://www.balena.io/docs/reference/balena-cli/#support-faq-and-troubleshooting

```

# Steps to Reproduce the Problem

I got this issue on my Native Ubuntu such as on a fresh installed VM.
I use balena-cli with openBalena
Dont hesitate to ask if you need any further Information.

1. Login
2. `balena-cli/balena preload <Image> --debug`

**Comment by pdcastro** [comment]
@HenrikWittemeier, thank you for reporting this issue. The error message sounds like a bug, but so far I have not been able to reproduce it. You mentioned a freshly installed VM, and it's been a while I wanted to put together some basic openBalena provisioning scripts for the purpose of debugging issues like this one. So I've written a couple of scripts and a README:

https://github.com/pdcastro/ob_multipass

Still somewhat experimental, but `multipass` worked well on my Mac. After provisioning the two VMs as described in the [README](https://github.com/pdcastro/ob_multipass#readme), I was able to successfully preload an image with the following commands on the `'localmachine'` VM:

```sh
$ multipass transfer ~/Downloads/balena-cloud-test-rpi-raspberrypi3-2.72.0+rev1-dev-v12.3.5.img.zip localmachine:
$ multipass exec localmachine bash
```
```sh
$ balena login --credentials -e "${OB_EMAIL}" -p "${OB_PASSWORD}"
$ balena app create myApp --type raspberrypi3
$ git clone https://github.com/balena-io-examples/balena-idling.git
$ balena deploy myApp --build -s ~/balena-idling
$ unzip balena-cloud-test-rpi-raspberrypi3-2.72.0+rev1-dev-v12.3.5.img.zip
$ balena preload balena-cloud-test-rpi-raspberrypi3-2.72.0+rev1-dev-v12.3.5.img -a myApp -c current --add-certificate /usr/local/share/ca-certificates/openbalena.crt --debug
```

Perhaps you could try to compare your setup with the above, for us to try and identify where the problem is. By the way, how did you install Docker? In my provisioning script, I did simply `sudo apt-get install docker.io`.

**Comment by HenrikWittemeier** [comment]
@pdcastro Ive tried your script for the `'localmachine'` VM. When trying to preload i got the same error as mentioned above. 
When i try to preload an application from **BalenaCloud** it works fine.
So i think the problem is not my installation of Balena-Cli, but something according to openBalena.
I'll check my openBalena installation!

**Comment by openedhardware** [comment]
Same issue with me - https://forums.balena.io/t/openbalena-preload-fails/311299

@HenrikWittemeier 
Did you figure out what was wrong with your openBalena server?

**Comment by HenrikWittemeier** [comment]
Hi @openedhardware,
at the Moment i have no idea where to start looking, so i have not figured out the problem.

**Comment by pdcastro** [comment]
Connects-to: balena-io-modules/balena-preload/issues/243

**Comment by pdcastro** [comment]
@HenrikWittemeier, [it has been reported](https://github.com/balena-io-modules/balena-preload/issues/243#issuecomment-843631779) that CLI v12.44.16 or later fixed this issue. I can't quite tell because I was not able to reproduce it in the first place. When / if you can, let us know if updating the CLI to that version makes a difference for you as well. Thanks!

**Comment by HenrikWittemeier** [comment]
Now everything is working fine :+1:

**Comment by pdcastro** [comment]
Fixed in CLI release v12.44.16 or later (pull requests #2269 and balena-io-modules/balena-preload/pull/244). Thanks @HenrikWittemeier for confirming it. 👍
